File: Sample/BasicSample/sample_cpp/Input/Joystick.cpp

#include <Altseed.h>

/**
@brief	Joystickのサンプル。ジョイスティックによる入力を取得しています。
*/
void Joystick()
{

	// Altseedを初期化する。
	asd::Engine::Initialize(asd::ToAString("Joystick").c_str(), 640, 480, asd::EngineOption());

	auto font = asd::Engine::GetGraphics()->CreateDynamicFont(asd::ToAString("").c_str(), 25, asd::Color(255, 255, 255, 255), 1, asd::Color(0, 0, 0, 255));

	auto buttonsStateText = std::make_shared<asd::TextObject2D>();
	buttonsStateText->SetPosition(asd::Vector2DF(10, 10));
	buttonsStateText->SetFont(font);

	asd::Engine::AddObject2D(buttonsStateText);

	// Altseedのウインドウが閉じられていないか確認する。
	while (asd::Engine::DoEvents())
	{
		asd::astring displayStr = asd::ToAString("");

		// ジョイスティックが最低一つ接続されているかどうかを確認する。

		// 一つも接続されていない場合は終了する。
		if (!asd::Engine::GetJoystickContainer()->GetIsPresentAt(0))
		{
			displayStr += asd::ToAString("No joysticks are connected.\n");
		}
		else
		{
			// 一つ目のジョイスティックの0番目のボタンの入力状態を表示する
			asd::Joystick* joystick = asd::Engine::GetJoystickContainer()->GetJoystickAt(0);

			for (int buttonIndex = 0; buttonIndex < joystick->GetButtonsCount(); ++buttonIndex)
			{
				switch (joystick->GetButtonState(buttonIndex))
				{
				case asd::JoystickButtonState::Free: //前フレームと本フレームで非押下
					displayStr += asd::ToAString(("Button " + std::to_string(buttonIndex) + ": Free").c_str());
					break;
				case asd::JoystickButtonState::Hold: //前フレームと本フレームで押下
					displayStr += asd::ToAString(("Button " + std::to_string(buttonIndex) + ": Hold").c_str());
					break;
				case asd::JoystickButtonState::Release: //前フレームで押下、本フレームで非押下
					displayStr += asd::ToAString(("Button " + std::to_string(buttonIndex) + ": Release").c_str());
					break;
				case asd::JoystickButtonState::Push: //前フレームで非押下、本フレームで押下
					displayStr += asd::ToAString(("Button " + std::to_string(buttonIndex) + ": Push").c_str());
					break;
				}

				displayStr += asd::ToAString("\n");
			}

		}
		
		buttonsStateText->SetText(displayStr.c_str());

		// Altseedを更新する。
		asd::Engine::Update();
	}

	//Altseedの終了処理をする。
	asd::Engine::Terminate();

	return;
}

File: Sample/BasicSample/sample_cs/Input/Joystick.cs

/// <summary>
/// Joystickのサンプル。ジョイスティックによる入力を取得しています。
/// </summary>
class Joystick : ISample
{
	public void Run()
	{
		// Altseedを初期化する。
		var initialized = asd.Engine.Initialize("Joystick", 640, 480, new asd.EngineOption());

		asd.JoystickContainer container = asd.Engine.JoystickContainer;

		// 0番目のジョイスティックが接続されているか否かを確かめる
		if (container.GetIsPresentAt(0))
		{
			// 一つも接続されていない場合は終了する
			System.Console.WriteLine("No joystick are connected.");
			asd.Engine.Terminate();
			return;
		}

		// Altseedのウインドウが閉じられていないか確認する。
		while (asd.Engine.DoEvents())
		{
			// 一つ目のジョイスティックの0番目のボタンの入力状態を表示する
			switch (container.GetJoystickAt(0).GetButtonState(0))
			{
				case asd.JoystickButtonState.Free: //前フレーム、現フレーム共に非押下。
					System.Console.WriteLine("Free");
					break;
				case asd.JoystickButtonState.Hold: //前フレーム、現フレーム共に押下。
					System.Console.WriteLine("Hold");
					break;
				case asd.JoystickButtonState.Release: //前フレームで押下、現フレームで非押下。
					System.Console.WriteLine("Release");
					break;
				case asd.JoystickButtonState.Push: //前フレームで非押下、現フレームで押下。
					System.Console.WriteLine("Push");
					break;
			}

			// Altseedを更新する。
			asd.Engine.Update();
		}
		//Altseedの終了処理をする。
		asd.Engine.Terminate();
	}
}

